Convert attribute decorated POCO's to Points

joakimhew opened the pull request:
This feature adds an extension method allowing users to convert attribute decorated POCO's following certain rules into [Points](https://github.com/pootzko/InfluxData.Net/blob/develop/InfluxData.Net.InfluxDb/Models/Point.cs).

The following attributes must be present in the POCO:

| Attribute name   |  Mandatory?  |  Maximum in one type  |
|-------------------|--------------|---------------------------|
| `[Measurement]` |    **Yes**    |  1                                   |
| `[Timestamp]`     |    **No**     |  1                                  |
| `[Tag]`                 |    **No**     |  ∞                                  |
| `[Field]`               |    **Yes**     |  ∞                                  |

Example: 

**Please notice how one of the tag and field names are explicitly set using the attribute and `name` argument**

```csharp
public class MyType 
{
     [Timestamp]
     public DateTime Time { get; set; }

     [Measurement]
     public string MyMeasurement { get; set; }

     [Tag]
     public string Tag1 { get; set; }

     [Tag(name: "newtagname")] // <- Explicitly set tag name
     public string Tag2 { get; set; }

     [Field]
     public string Field1 { get; set; }

     [Field(name: "newfieldname")] // <- Explicitly set field name
     public bool Field2 { get; set; }

     [Field]
     public float Field2 { get; set; }
}

MyType myType = new MyType
{
    Time = DateTime.Now,
    Measurement = "TestMeasurement",
    Tag1 = "FirstTag",
    Tag2 = "SecondTag",
    Field1 = "FirstField",
    Field2 = True,
    Field3 = 2324
};

var point = myType.ToPoint();
```

This will produce the following point object:

```javascript
point = 
{
    "Timestamp" : "2017-10-17 10:18:00",
    "Name" : "TestMeasurement",
    "Tags" : [
          "Tag1" : "FirstTag",
          "newtagname`" : "SecondTag" // <- explicitly set tag name
    ],
    "Fields" : [
          "Field1" : "FirstField",
          "newfieldname" : True, // <- explicitly set field name
          "Field3" : 2324
     ]
}
```

The following validations are implemented and will return detailed exceptions:
- [Makes sure tags are only of type `string`](https://github.com/pootzko/InfluxData.Net/pull/64/files#diff-0d1bebcc18d7851815026e2ae60ffe7dR134)
- [Makes sure fields are of type `primitive` or `string`](https://github.com/pootzko/InfluxData.Net/pull/64/files#diff-0d1bebcc18d7851815026e2ae60ffe7dR159)
- [Makes sure there is **EXACTLY** one `[Measurement]` present](https://github.com/pootzko/InfluxData.Net/pull/64/files#diff-0d1bebcc18d7851815026e2ae60ffe7dR101)
- [Makes sure the property decorated with `[Measurement]` is of type `string`](https://github.com/pootzko/InfluxData.Net/pull/64/files#diff-0d1bebcc18d7851815026e2ae60ffe7dR115)
- [If used, make sure there is **EXACTLY** one `[Timestamp]` present](https://github.com/pootzko/InfluxData.Net/pull/64/files#diff-0d1bebcc18d7851815026e2ae60ffe7dR73)
- [Makes sure the property decorated with `[Timestamp]` is of type `DateTime`](https://github.com/pootzko/InfluxData.Net/pull/64/files#diff-0d1bebcc18d7851815026e2ae60ffe7dR79)

Commit:
Add attributes

Commit:
Add measurement attribute

Commit:
Move measurement attribute into correct namespace

Commit:
Add MissingExpectedAttribyteException

Commit:
Add point extension for converting decorated types into 'Point'

Commit:
Add unit tests for point extensions

tihomir-kit commented:
Lol! Dude.. this is a sick feature. Just awesome. :)

I'll try to publish it over the weekend.

Thanks!

Commit:
Add option to override property name in field and tag attribute

joakimhew commented:
@pootzko Glad I can help! I just added the option to explicitly set the names for tags and fields using a name argument in the attributes. Updated my original comment.

I found out at work today that this would be nice to have so figured I could make a proper implementation!

Commit:
Remove unecessary initialization of Tags and Fields dictionary

Commit:
Minor cleanup and misplaced if statement

Commit:
Add test for POCO without any field attributes

joakimhew commented:
Have you had the chance to look more into this 😄 ?

tihomir-kit commented:
Hey man, sorry for the delay. I haven't yet, I've been very busy the last 2 weeks. I'll try to look at it this weekend.
